crates/core/tedge/src/cli/mqtt/mod.rs

```diff
@@ -1,41 +1,64 @@
 pub use self::cli::TEdgeMqttCli;
-pub use self::error::MqttError;
-use rumqttc::Client;
-use std::sync::atomic::AtomicBool;
-use std::sync::atomic::Ordering;
-use std::sync::Arc;
+use std::future::Future;
+use std::pin::Pin;
 use std::time::Duration;
+use tokio::signal::unix;
+use tokio::time;
 
 mod cli;
-mod error;
 mod publish;
 mod subscribe;
 
 const MAX_PACKET_SIZE: usize = 268435455; // 256 MB
 
-fn disconnect_if_interrupted(client: Client, timeout: Option<Duration>) -> Arc<AtomicBool> {
-    let interrupted = Arc::new(AtomicBool::new(false));
-    for signal in signal_hook::consts::TERM_SIGNALS {
-        let client = client.clone();
-        let interrupted = interrupted.clone();
-        unsafe {
-            let _ = signal_hook::low_level::register(*signal, move || {
-                eprintln!("INFO: Interrupted");
-                interrupted.store(true, Ordering::Relaxed);
-                let _ = client.disconnect();
-            });
-        }
-    }
+#[derive(Copy, Clone, Debug)]
+enum Interruption {
+    Timeout,
+    Interrupted,
+}
+
+struct TermSignals {
+    signals: Option<unix::Signal>,
+    timeout: Option<Pin<Box<time::Sleep>>>,
+}
 
-    if let Some(timeout) = timeout {
-        let timeout_elapsed = interrupted.clone();
-        std::thread::spawn(move || {
-            std::thread::sleep(timeout);
-            eprintln!("INFO: Timeout");
-            timeout_elapsed.store(true, Ordering::Relaxed);
-            let _ = client.disconnect();
-        });
+impl TermSignals {
+    fn new(timeout: Option<Duration>) -> TermSignals {
+        let signals = unix::signal(unix::SignalKind::interrupt())
```

Comment:
System test "Subscribe Connection Closed On Interruption" is failing because I forgot to also handle [`quit`](https://docs.rs/tokio/latest/tokio/signal/unix/struct.SignalKind.html#method.quit) and  [`terminate`](https://docs.rs/tokio/latest/tokio/signal/unix/struct.SignalKind.html#method.terminate).

Reply:
Fixed by https://github.com/thin-edge/thin-edge.io/pull/3466/commits/7fe45c0784a62e7ca75b537c3b0a2a858fa9dd4d.

This is working but still too complicated. I'm still looking for something simpler.

Reply:
Simplified: https://github.com/thin-edge/thin-edge.io/pull/3466/commits/efec11021d76289ab605109b25e793fe9fd6abb0